### PR TITLE
Check for unknown object keys before allocating SubscriptFunctions

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
2nd attempt to https://github.com/crate/crate/pull/14485 (to address https://github.com/crate/crate/pull/14371#discussion_r1262735707).

Fixes:
```
cr> create table t (o object);                                                                                                                    
CREATE OK, 1 row affected  (1.624 sec)

cr> select o['unknown'] from t;                                                                                                                   
ColumnUnknownException[Column o['unknown'] unknown]

cr> select o['unknown'] from t as t2; -- should throw like above                                                                                                             
+--------------+
| o['unknown'] |
+--------------+
+--------------+
SELECT 0 rows in set (0.003 sec)
```

limiting the scope, considering `References` as the bases of subscript expressions and string literals as the indexes.

The approach is to add more checks before allocating `SubscriptFunctions` when analyzing `SubscriptExpressions`.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
Cases like below would not be fixed by this PR, as the evaluation of functions happen after the analysis of subscript expressions:
```
select obj[concat('unknown', 'key')] from t; -- key is a function

select o[1]['o2']::array(object)[1]['u'] from t; -- base is a function
```

**Comment by jeeminso** [comment]
I have not added any tests yet but could you have a rough look @mfussenegger @matriv ? Thank you

File: server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -707,9 +709,16 @@ protected Symbol visitSubscriptExpression(SubscriptExpression node, ExpressionAn
                                                                  List.of(),
                                                                  operation,
                                                                  context.errorOnUnknownObjectKey());
-                        if (base instanceof Reference) {
+                        if (base instanceof Reference ref && ref.columnPolicy() != ColumnPolicy.DYNAMIC ||
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:712]: I think with `object(ignored)` this also shouldn't raise?

File: server/src/main/java/io/crate/analyze/relations/UnionSelect.java [hunk 2]
```diff
@@ -53,17 +55,7 @@ public UnionSelect(AnalyzedRelation left, AnalyzedRelation right, boolean isDist
...
-        for (int i = 0; i < fieldsFromLeft.size(); i++) {
-            Symbol field = fieldsFromLeft.get(i);
-            Symbol rightField = right.outputs().get(i);
-            DataType<?> type = field.valueType().precedes(rightField.valueType())
-                ? field.valueType()
-                : rightField.valueType();
-            outputs.add(new ScopedSymbol(name, Symbols.pathFromSymbol(field), type));
-        }
-        this.outputs = List.copyOf(outputs);
+        this.outputs = List.copyOf(unionizeOutputs(left.outputs(), right.outputs()));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/relations/UnionSelect.java:58]: Why do we need this? Bias for the left relation for union should be fine. No?

Even if not, this seems to be unrelated to the main change of this PR, which afaik is to address this:

```
cr> select o['unknown'] from t;                                                                                                                   
ColumnUnknownException[Column o['unknown'] unknown]

cr> select o['unknown'] from t as t2; -- should throw like above            
```

Comment (matriv) [reply server/src/main/java/io/crate/analyze/relations/UnionSelect.java:58]: Could it be because of a case where `o['uknown']` exists in left relation but not in right?

Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/relations/UnionSelect.java:58]: I am adding a step to validate the object key exists in the `ObjectType` before allocating subscript functions which breaks analysis of `union` if selecting sub-column of object from right table.

```
create table v1 (obj object as (a int));
create table v2 (obj object as (b int));
select obj['a'], obj['b'] from (select obj from v1 union all select obj from v2) v; -- cannot select 'b' without merging 'obj'
```

File: server/src/main/java/io/crate/expression/symbol/ScopedSymbol.java [hunk 3]
```diff
@@ -54,11 +56,13 @@ public final class ScopedSymbol implements Symbol {
     private final RelationName relation;
     private final ColumnIdent column;
     private final DataType<?> dataType;
+    private final ColumnPolicy columnPolicy;
 
-    public ScopedSymbol(RelationName relation, ColumnIdent column, DataType<?> dataType) {
+    public ScopedSymbol(RelationName relation, ColumnIdent column, Symbol symbol) {
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/symbol/ScopedSymbol.java]: Thank you, please have a look at https://github.com/crate/crate/pull/14593/commits/21479ebd3d175e0391533957aa132ea611318b8a. If the fix is reasonable, I can move on to removing `ColumnPolicy` from `References`. Could you have a look @matriv ?

[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/io/crate/expression/symbol/ScopedSymbol.java]: I think it looks good, though, could you please explain what's the issue that this approach is solving?

[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/symbol/ScopedSymbol.java]: Thank you @matriv , it is needed to be able to access the ColumnPolicy from ExpressionAnalyzer: https://github.com/crate/crate/blob/17b8d6ac8a56122aaa02d1f56dcb0f006bbea60b/server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java#L712-L717

File: server/src/main/java/io/crate/expression/reference/doc/lucene/LuceneReferenceResolver.java [hunk 4]
```diff
@@ -159,7 +158,7 @@ private static LuceneCollectorExpression<?> typeSpecializedExpression(final Fiel
         final String fqn = ref.column().fqn();
         final MappedFieldType fieldType = fieldTypeLookup.get(fqn);
         if (fieldType == null) {
-            return NO_FIELD_TYPES_IDS.contains(unnest(ref.valueType()).id()) || isIgnoredDynamicReference(ref)
+            return NO_FIELD_TYPES_IDS.contains(unnest(ref.valueType()).id()) || ref.symbolType() == SymbolType.IGNORED_REFERENCE
```
Comment (jeeminso) [inline server/src/main/java/io/crate/expression/reference/doc/lucene/LuceneReferenceResolver.java:161]: This lead `IgnoredReference`.

File: server/src/test/java/io/crate/analyze/UnionAnalyzerTest.java [hunk 5]
```diff
@@ -197,9 +197,9 @@ public void testUnionObjectTypesWithSubColumnsOfSameNameButDifferentInnerTypes()
             ).build();
 
         UnionSelect union = analyze("select obj from v1 union all select obj from v2");
-        ObjectType expectedType = ObjectType.builder().setInnerType("col", DataTypes.STRING).build();
+        ObjectType expectedType = ObjectType.builder().setInnerType("col", DataTypes.INTEGER).build();
         Asserts.assertThat(union.outputs())
-            .as("Output type is object(col::text) because text has higher precedence than int")
+            .as("Output type is object(col::int) because int has higher precedence than text")
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/test/java/io/crate/analyze/UnionAnalyzerTest.java]: Caused by error in one of the commits, fix.

**Comment by jeeminso** [comment]
Please review this @mfussenegger each commits should be self contained, Thank you

File: server/src/main/java/io/crate/analyze/relations/UnionSelect.java [hunk 6]
```diff
@@ -109,4 +102,42 @@ public String toString() {
...
+        ArrayList<ScopedSymbol> outputs = new ArrayList<>(left.size());
+        for (int i = 0; i < left.size(); i++) {
+            outputs.add(unionizeSymbol(left.get(i), right.get(i)));
+        }
+        return outputs;
+    }
+
+    private ScopedSymbol unionizeSymbol(Symbol left, Symbol right) {
+        return new ScopedSymbol(name, Symbols.pathFromSymbol(left), unionizeTypes(left.valueType(), right.valueType()));
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/relations/UnionSelect.java:116]: I'd keep this inlined in the ctor. It's not re-used and hardly worth breaking out.

File: server/src/main/java/io/crate/analyze/relations/UnionSelect.java [hunk 7]
```diff
@@ -109,4 +102,42 @@ public String toString() {
...
+    private ObjectType unionizeObjectType(ObjectType left, ObjectType right) {
+        ObjectType.Builder unionizedObjectBuilder = ObjectType.builder();
+        for (var e : left.innerTypes().entrySet()) {
+            unionizedObjectBuilder.setInnerType(e.getKey(), e.getValue());
+        }
+        for (var e : right.innerTypes().entrySet()) {
+            unionizedObjectBuilder.mergeInnerType(e.getKey(), e.getValue(), this::unionizeTypes);
+        }
+        return unionizedObjectBuilder.build();
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/relations/UnionSelect.java:142]: I think this could go into the `ObjectType` as `ObjectType merge(ObjectType other)`.
Maybe it is even worth to have a `DataTypes.merge(left, right)` to replace the `unionizeTypes`

I'd avoid the `unionize` term. It's a new term we don't use so far and has a different meaning from what it's being used for. Stick to `merge` which is established already.

File: server/src/main/java/io/crate/types/DataTypes.java [hunk 8]
```diff
@@ -486,7 +487,10 @@ public static DataType<?> ofNameOrNull(String typeName) {
         entry("ip", DataTypes.IP),
         entry("geo_point", DataTypes.GEO_POINT),
         entry("geo_shape", DataTypes.GEO_SHAPE),
-        entry("object", UNTYPED_OBJECT),
+        // the mapping name is "object" but column policies are prepended in order to differentiate empty object types.
+        entry(ColumnPolicy.DYNAMIC.lowerCaseName() + "_object", ObjectType.DEFAULT_EMPTY_OBJECT),
+        entry(ColumnPolicy.IGNORED.lowerCaseName() + "_object", ObjectType.IGNORED_EMPTY_OBJECT),
+        entry(ColumnPolicy.STRICT.lowerCaseName() + "_object", ObjectType.STRICT_EMPTY_OBJECT),
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/types/DataTypes.java:493]: It was for `DocIndexMetadata.getColumnDataType` where it was used to select the empty object types now including column policies. I have created `ObjectType.of(ColumnPolicy)` instead.

File: server/src/main/java/io/crate/analyze/relations/UnionSelect.java (hunk 7 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/analyze/relations/UnionSelect.java]: Thank you, done

**Comment by jeeminso** [comment]
Thanks for reviewing @mfussenegger , please have another look.

File: server/src/main/java/io/crate/expression/symbol/DynamicReference.java [hunk 9]
```diff
@@ -21,29 +21,36 @@
...
+    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, int position, ColumnPolicy columnPolicy) {
         super(ident, granularity, DataTypes.UNDEFINED, position, null);
+        this.columnPolicy = columnPolicy;
     }
 
-    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, ColumnPolicy columnPolicy, int position) {
-        super(ident, granularity, DataTypes.UNDEFINED, columnPolicy, IndexType.PLAIN, true, false, position, null);
+    public DynamicReference(StreamInput in) throws IOException {
+        super(in);
+        columnPolicy = in.readEnum(ColumnPolicy.class);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/symbol/DynamicReference.java:47]: This would need version bwc handling,  but can't we derive this from the type?

File: server/src/main/java/io/crate/types/DataTypes.java [hunk 10]
```diff
@@ -610,6 +610,19 @@ public static boolean isCompatibleType(List<DataType<?>> left, List<DataType<?>>
         return true;
     }
 
+    public static DataType<?> merge(DataType<?> leftType, DataType<?> rightType) {
+        DataType<?> type;
+        if (leftType.id() == ObjectType.ID) {
+            type = ObjectType.merge((ObjectType) leftType, (ObjectType) rightType);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/DataTypes.java:616]: This fails if the right type isn't an object

File: server/src/main/java/io/crate/types/DataTypes.java [hunk 11]
```diff
@@ -610,6 +610,19 @@ public static boolean isCompatibleType(List<DataType<?>> left, List<DataType<?>>
...
+        } else if (leftType.id() == ArrayType.ID) {
+            type = new ArrayType<>(
+                merge(((ArrayType<?>) leftType).innerType(), ((ArrayType<?>) rightType).innerType()));
+        } else {
+            type = leftType.isConvertableTo(rightType, false) ? rightType : leftType;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/DataTypes.java:621]: Shouldn't this also check the precedence?

File: server/src/main/java/io/crate/types/ObjectType.java [hunk 12]
```diff
@@ -78,17 +81,40 @@ public ValueIndexer<Map<String, Object>> valueIndexer(RelationName table,
         }
     };
 
+    public static ObjectType of(ColumnPolicy columnPolicy) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/types/ObjectType.java:84]: I'm not sure if we want this.  Shouldn't most objectType creation code go through the builder to also set inner types? 

We probably shouldn't make it too convenient to use empty instances. Same goes for the constants. Do we really need all 3?

**Comment by mfussenegger** [comment]
Also cc @seut - would be good to have your input regarding the move of ColumnPolicy from Reference to the ObjectType.

File: server/src/main/java/io/crate/expression/symbol/DynamicReference.java (hunk 9 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/symbol/DynamicReference.java:47]: Thank you, adding the version checks. I think the type is fixed to `Undefined` so cannot be derived.

File: server/src/main/java/io/crate/types/DataTypes.java (hunk 10 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/types/DataTypes.java:616]: Thanks for pointing out, I just blindly took it out of UnionSelect.

File: server/src/main/java/io/crate/types/DataTypes.java (hunk 11 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/types/DataTypes.java:621]: Thanks, this helped me realize a bug that slipped my mind. I am working on a [PR](https://github.com/crate/crate/pull/14723) for it and move `union` related commit to it.

File: server/src/main/java/io/crate/expression/symbol/DynamicReference.java [hunk 13]
```diff
@@ -21,29 +21,44 @@
...
 
-    public DynamicReference(StreamInput in) throws IOException {
-        super(in);
-    }
+    private final ColumnPolicy columnPolicy;
 
-    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, int position) {
+    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, int position, ColumnPolicy columnPolicy) {
         super(ident, granularity, DataTypes.UNDEFINED, position, null);
+        this.columnPolicy = columnPolicy;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/expression/symbol/DynamicReference.java:44]: I'm a bit confused. Part of the motivation for moving `ColumnPolicy` to the `ObjectType` was that only objects have a policy, and regular columns don't.

If we need to add a columnPolicy to the `DynamicReference`, that's apparently not the case and we may be better off keeping it on `(Simple)Reference` ?

What's a scenario where the columnPolicy on a non-object column for dynamic matters/is different?

**Review by mfussenegger — COMMENTED** [review]
I took a closer look at this and think we should put this on hold for now, create an issue for later proper planned feature work.

The issue here isn't just related to aliases, but we've generally a mix of fail-eager and fail-lazy when it comes to subscript expressions.

For example:

```
cr> select o['x'] from (select {y=1} o from generate_series(0, 1) t) t1
ColumnUnknownException[The object `{y=1}` does not contain the key `x`]
```

But:

```
cr> select o['x'] from (select {y=1} o from generate_series(0, -1) t) t1
+--------+
| o['x'] |
+--------+
+--------+
```

The reason is that we defer the error handling to runtime (via `SubscriptFunction#evaluate`) when we can't directly resolve a subscript:

https://github.com/crate/crate/blob/608a7f2b1fcd61842904a4a94d119d7a954f3e65/server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java#L706-L720


One option could be to look into making sure an object-type always has inner-types, and then use that information to decide eagerly.

E.g. change https://github.com/crate/crate/blob/608a7f2b1fcd61842904a4a94d119d7a954f3e65/server/src/main/java/io/crate/expression/symbol/Literal.java#L217


To:

```
public static Literal<Map<String, Object>> of(Map<String, Object> value) {
    Builder builder = ObjectType.builder();
    for (var entry : value.entrySet()) {
        String fName = entry.getKey();
        Object fValue = entry.getValue();
        builder.setInnerType(fName, DataTypes.guessType(fValue));
    }
    return new Literal<>(builder.build(), value);
}
```

But we'd have to go through all components dealing with objects and make sure they provide inner types.


---

There may be other options. In any case we should try to come up with a proper design that simplifies things instead of adding more patchwork.

**Comment by mfussenegger** [comment]
See https://github.com/crate/crate/issues/14828
